### ISSUE TITLE
ImportError: dlopen: cannot load any more object with static TLS

### ISSUE THREAD

**Issue by Bobo22331** [wangyibin/CPhasing#1]
When I run the cphasing using Hi-C data, it occurs some errors:
```
Traceback (most recent call last):
  File "/home/cuixb/tools/biosoft/conda3/envs/cphasing/bin/cphasing", line 7, in <module>
    sys.exit(cli())
             ^^^^^
  File "/home/cuixb/tools/biosoft/conda3/envs/cphasing/lib/python3.12/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/cuixb/tools/biosoft/conda3/envs/cphasing/lib/python3.12/site-packages/rich_click/rich_command.py", line 126, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/home/cuixb/tools/biosoft/conda3/envs/cphasing/lib/python3.12/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/cuixb/tools/biosoft/conda3/envs/cphasing/lib/python3.12/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/cuixb/tools/biosoft/conda3/envs/cphasing/lib/python3.12/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/cuixb/tools/biosoft/CPhasing/cphasing/cli.py", line 995, in pipeline
    run(fasta,
    ^^^^^^^^^^
  File "/home/cuixb/tools/biosoft/CPhasing/cphasing/pipeline/pipeline.py", line 108, in run
    from ..chimeric import run as chimeric_run
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/cuixb/tools/biosoft/CPhasing/cphasing/chimeric.py", line 28, in <module>
    from sklearn.metrics import mean_squared_error
  File "/home/cuixb/tools/biosoft/conda3/envs/cphasing/lib/python3.12/site-packages/sklearn/__init__.py", line 74, in <module>
    from .utils._show_versions import show_versions  # noqa: E402
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/cuixb/tools/biosoft/conda3/envs/cphasing/lib/python3.12/site-packages/sklearn/utils/_show_versions.py", line 16, in <module>
    from ._openmp_helpers import _openmp_parallelism_enabled
ImportError: dlopen: cannot load any more object with static TLS
```
My command:
```
cphasing pipeline -f curated.fasta -hic1 E4-P8-2_L1_335X35.R1.fastq.gz -hic2 E4-P8-2_L1_335X35.R2.fastq.gz -t 20 -n 10
```
Thanks a lot!

**Comment by wangyibin** [comment]
Sorry for the late reply.

Can you show me the version of `sklearn`? 
Or you can try update the CPhasing to the latest version, I have reconstructed the environment since `v0.2.0.r270`.

**Comment by Bobo22331** [comment]
This is my scikit-learn version:
```
scikit-learn              1.6.0                    pypi_0    pypi
```

**Comment by wangyibin** [comment]
Hi,
I tested it on multiple machines, and it didn't cause the same error, so I don't know how to fix it right now.

However, in the latest CPhasing `v0.2.0.r275`, the `pipeline` avoids importing `sklearn` when it is not in use, and you can avoid this problem for the time being when you do not specify the `--chimeric-correct` parameter.

**Comment by Bobo22331** [comment]
Many thanks. I have updated to 0.2.0.r275, but other error occurs:
```
[20:42:49] INFO     C-Phasing version: 0.2.0.r275                pipeline.py:111
           INFO     Pipeline is started on 2024-12-13.           pipeline.py:113
           INFO     Mode is set to `haploid` because of the      pipeline.py:128
                    second `n` is not specified.                                
           INFO     Working on existing directory:               pipeline.py:133
                    `cphasing_output`                                           
           INFO     The mode is `haploid`, skip step 1. alleles. pipeline.py:212
           INFO     #----------------------------------#         pipeline.py:325
                    #      Running step 0. mapper      #                        
                    #----------------------------------#                        
Traceback (most recent call last):
  File "/home/cuixb/tools/biosoft/conda3/envs/cphasing/bin/cphasing", line 7, in <module>
    sys.exit(cli())
             ^^^^^
  File "/home/cuixb/tools/biosoft/conda3/envs/cphasing/lib/python3.12/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/cuixb/tools/biosoft/conda3/envs/cphasing/lib/python3.12/site-packages/rich_click/rich_command.py", line 126, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/home/cuixb/tools/biosoft/conda3/envs/cphasing/lib/python3.12/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/cuixb/tools/biosoft/conda3/envs/cphasing/lib/python3.12/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/cuixb/tools/biosoft/conda3/envs/cphasing/lib/python3.12/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/cuixb/tools/biosoft/conda3/envs/cphasing/lib/python3.12/site-packages/cphasing/cli.py", line 995, in pipeline
    run(fasta, 
    ^^^^^^^^^^
  File "/home/cuixb/tools/biosoft/conda3/envs/cphasing/lib/python3.12/site-packages/cphasing/pipeline/pipeline.py", line 355, in run
    hic_mapper.main(args=[
  File "/home/cuixb/tools/biosoft/conda3/envs/cphasing/lib/python3.12/site-packages/rich_click/rich_command.py", line 126, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/home/cuixb/tools/biosoft/conda3/envs/cphasing/lib/python3.12/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/cuixb/tools/biosoft/conda3/envs/cphasing/lib/python3.12/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/cuixb/tools/biosoft/conda3/envs/cphasing/lib/python3.12/site-packages/cphasing/hic/cli.py", line 182, in mapper
    cm = ChromapMapper(reference, r1, r2, 
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/cuixb/tools/biosoft/conda3/envs/cphasing/lib/python3.12/site-packages/cphasing/mapper.py", line 255, in __init__
    raise ValueError(f"{self._path}: command not found")
ValueError: _chromap: command not found
```
but I have installed the chromap tool.
```
(cphasing) [cuixb@yls run1]$ which chromap
~/tools/biosoft/conda3/envs/cphasing/bin/chromap
```

**Comment by Bobo22331** [comment]
An error occurs with step 3 hyperpartition, here is the log information:
```
[21:51:53] INFO     C-Phasing version: 0.2.0.r275                pipeline.py:111
           INFO     Pipeline is started on 2024-12-13.           pipeline.py:113
           INFO     Mode is set to `haploid` because of the      pipeline.py:128
                    second `n` is not specified.                                
           INFO     Working on existing directory:               pipeline.py:133
                    `cphasing_output`                                           
           INFO     The mode is `haploid`, skip step 1. alleles. pipeline.py:212
           INFO     #----------------------------------#         pipeline.py:325
                    #      Running step 0. mapper      #                        
                    #----------------------------------#                        
           INFO     Running command:                             utilities.py:97
           INFO             cphasing-rs chromsizes curated.fasta utilities.py:98
                    -o curated.contigsizes                                      
[21:52:11] INFO     Running command:                             utilities.py:97
           INFO             _chromap -t 20 -k 17 -w 7 -i -r      utilities.py:98
                    curated.fasta -o curated.index                              
[21:56:41] INFO     Running command:                             utilities.py:97
           INFO             _chromap -t 20 --preset hic -k 17 -w utilities.py:98
                    7 -q 0 -x curated.index                                     
                    --remove-pcr-duplicates -r curated.fasta -1                 
                    /raid9/cuixb/analysis/Brat2t/02.scaffolding/                
                    purged_haplotigs_scaffolding/purge_haplotigs                
                    _version/cphasing/run1/E4-P8-2_L1_335X35.R1.                
                    fastq.gz -2                                                 
                    /raid9/cuixb/analysis/Brat2t/02.scaffolding/                
                    purged_haplotigs_scaffolding/purge_haplotigs                
                    _version/cphasing/run1/E4-P8-2_L1_335X35.R2.                
                    fastq.gz -o E4-P8-2_L1_335X35.R1.pairs                      
[23:47:06] INFO     Done.                                          mapper.py:323
           INFO     Running command:                             utilities.py:97
           INFO             pigz -f -p 20                        utilities.py:98
                    E4-P8-2_L1_335X35.R1.pairs                                  
[23:47:36] INFO     Done.                                          mapper.py:329
           INFO                                                  pipeline.py:734
                    #----------------------------------#                        
                    #       Running step 2. prepare    #                        
                    #----------------------------------#                        
[23:47:37] INFO     Running command:                             utilities.py:97
           INFO             cphasing-rs count_re                 utilities.py:98
                    ../curated.fasta -o                                         
                    E4-P8-2_L1_335X35.R1.counts_AAGCTT.txt -p                   
                    AAGCTT                                                      
[23:47:44] INFO     Running command:                             utilities.py:97
           INFO             cphasing-rs pairs2clm                utilities.py:98
                    ../E4-P8-2_L1_335X35.R1.pairs.gz -c 1 -t 20                 
                    -o E4-P8-2_L1_335X35.R1.clm.gz -q 0                         
[23:49:42] INFO                                                  pipeline.py:833
                    #----------------------------------#                        
                    #  Running step 3. hyperpartition  #                        
                    #----------------------------------#                        
Traceback (most recent call last):
  File "/home/cuixb/tools/biosoft/CPhasing/bin/cphasing", line 7, in <module>
    sys.exit(cli())
             ^^^^^
  File "/home/cuixb/tools/biosoft/conda3/envs/cphasing/lib/python3.12/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/cuixb/tools/biosoft/conda3/envs/cphasing/lib/python3.12/site-packages/rich_click/rich_command.py", line 126, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/home/cuixb/tools/biosoft/conda3/envs/cphasing/lib/python3.12/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/cuixb/tools/biosoft/conda3/envs/cphasing/lib/python3.12/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/cuixb/tools/biosoft/conda3/envs/cphasing/lib/python3.12/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/cuixb/tools/biosoft/conda3/envs/cphasing/lib/python3.12/site-packages/cphasing/cli.py", line 995, in pipeline
    run(fasta, 
    ^^^^^^^^^^
  File "/home/cuixb/tools/biosoft/conda3/envs/cphasing/lib/python3.12/site-packages/cphasing/pipeline/pipeline.py", line 918, in run
    hyperpartition.main(args=hyperpartition_args,
  File "/home/cuixb/tools/biosoft/conda3/envs/cphasing/lib/python3.12/site-packages/rich_click/rich_command.py", line 126, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/home/cuixb/tools/biosoft/conda3/envs/cphasing/lib/python3.12/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/cuixb/tools/biosoft/conda3/envs/cphasing/lib/python3.12/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/cuixb/tools/biosoft/conda3/envs/cphasing/lib/python3.12/site-packages/cphasing/cli.py", line 3297, in hyperpartition
    from .hypergraph import HyperExtractor, Extractor
  File "/home/cuixb/tools/biosoft/conda3/envs/cphasing/lib/python3.12/site-packages/cphasing/hypergraph.py", line 26, in <module>
    from .algorithms.hypergraph import HyperEdges
  File "/home/cuixb/tools/biosoft/conda3/envs/cphasing/lib/python3.12/site-packages/cphasing/algorithms/hypergraph.py", line 9, in <module>
    import igraph as ig
  File "/home/cuixb/tools/biosoft/conda3/envs/cphasing/lib/python3.12/site-packages/igraph/__init__.py", line 25, in <module>
    from igraph._igraph import (
ImportError: dlopen: cannot load any more object with static TLS
```

**Comment by wangyibin** [comment]
Sorry for the late reply.
I found this problem mentioned in [this issue](https://github.com/scikit-learn/scikit-learn/issues/14485).
It can be fixed by updating the `glibc >=2.21`.

**Comment by Bobo22331** [comment]
Many thanks! Unfortunately I don't have access to update the glibc.
